Extend Transformers Trainer Class to Enable CPU AMP and Integrate Intel Extension for PyTorch

jianan-gu opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR supports the feature request https://github.com/huggingface/transformers/issues/17137



## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

jianan-gu committed:
init PR

jianan-gu committed:
fix import ipex

jianan-gu committed:
minor fix on bf16

jianan-gu committed:
refine optimizer

jianan-gu committed:
refine args notes

jianan-gu committed:
refine code

jianan-gu committed:
refine ipex optimize args

jianan-gu committed:
refine half_precision_backend

jgong5 commented:
cc @stas00

jianan-gu committed:
black format

jianan-gu committed:
isort format

jianan-gu committed:
isort format files

jianan-gu committed:
flake8 format

jianan-gu committed:
doc builder format

jianan-gu committed:
Merge branch 'huggingface:main' into Introduce_IPEX

stas00 commented on src/transformers/training_args.py:
```diff
@@ -876,10 +898,10 @@ def __post_init__(self):
             and (self.device.type != "cuda")
             and not (self.device.type == "xla" and "GPU_NUM_DEVICES" in os.environ)
-            and (self.fp16 or self.fp16_full_eval or self.bf16 or self.bf16_full_eval)
+            and (self.fp16 or self.fp16_full_eval)
```
is the bf16  code added earlier and removed from here really an equivalent? the new code you added earlier isn't checking all the same checks as this one. Specifically I'm not sure about xla checks.

jianan-gu replied:
Have made changes to make this checking between "device" and "precision" more equivalent.
Spliting the original checking on fp16 and bf16 with cuda device into two:
fp16 works on Cuda device;
bf16 works on Cuda and CPU device;
Thanks.

stas00 reviewed (changes requested):
Thank you for the detailed proposal and the implementation, @jianan-gu!

Other than the few comments/suggestions I left in the code the next step is we need tests that exercise this new feature.

The failing `run_tests_torch_and_flax` CI has a flakey test - please ignore.

jianan-gu committed:
refine codes

jianan-gu commented on src/transformers/trainer.py:
```diff
@@ -150,10 +152,16 @@
+    _is_native_cuda_amp_available = True
+
+if version.parse(torch.__version__) >= version.parse("1.10"):
+    _is_torch_generator_available = True
```
Have made changes to remove this duplicated line.

jianan-gu commented on src/transformers/trainer.py:
```diff
@@ -447,23 +455,35 @@ def __init__(
+                            args.half_precision_backend = "apex"
+                elif args.device == torch.device("cpu"):
+                    if args.fp16:
+                        raise ValueError("Tried to use `fp16` but is not supported on cpu")
```
Have made changes to refine this notice info.

stas00 replied:
typically, you can just hit "Commit suggestion" button and and all the work has been done for you already. and you don't have to comment if you're in agreement. will just save you a lot of time.

But of course feel free to do it in any fashion that works for you.

stas00 replied:
the good side-effect of hitting "Commit suggestion" is that it instantly resolves the conversation.

jianan-gu commented on src/transformers/trainer.py:
```diff
@@ -476,12 +496,20 @@ def __init__(
                 else:
                     self.scaler = torch.cuda.amp.GradScaler()
-            else:
+            elif args.half_precision_backend == "apex":
```
Have made changes to re-org this part accordingly.

jianan-gu commented on src/transformers/trainer.py:
```diff
@@ -2018,11 +2093,16 @@ def autocast_smart_context_manager(self):
+                ctx_manager = torch.cpu.amp.autocast(dtype=self.amp_dtype)
             else:
-                ctx_manager = autocast()
+                ctx_manager = torch.cpu.amp.autocast()
```
Totally agree, have made changes to refine this part accordingly (and the above  `()`), thanks.

stas00 commented on src/transformers/trainer.py:
```diff
@@ -937,6 +965,10 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
                 optimizer_kwargs.update(adam_kwargs)
             except ImportError:
                 raise ValueError("Trainer tried to instantiate bnb Adam8bit but bnb is not installed!")
+        elif args.optim == OptimizerNames.SGD:
+            optimizer_cls = torch.optim.SGD
+        elif args.optim == OptimizerNames.ADAGRAD:
+            optimizer_cls = torch.optim.Adagrad
```
Perhaps my comment wasn't clear. We typically try not to have all-in-one PRs that combine multiple features that don't depend on each other at once.

So this change is  welcome just in a separate PR if it makes sense?

jianan-gu commented on src/transformers/trainer.py:
```diff
@@ -1055,7 +1087,50 @@ def call_model_init(self, trial=None):
+                model = self.ipex_optimize_model(model, training, dtype=torch.bfloat16)
+            else:
+                model = self.ipex_optimize_model(model, training, dtype=torch.float32)
+        if self.args.jit_mode:
+            model = self.torch_jit_model(model, training, dataloader)
```
As mentioned in the https://github.com/huggingface/transformers/issues/17137 (Graph Optimization):
IPEX brings many benefits for model inference with its graph optimization, while the graph optimization is mostly based on PyTorch torchscript (--use_ipex + --jit_mode).

stas00 replied:
I totally understand. My comment wasn't against adding it but about how this is best done in our workflow. Therefore as I mentioned earlier, this is a separate PR please.

For example, you could do 2 PRs to handle the additions of the Optimizer bits and jit code - get those merged, then rebase this PR to use those features, and merge this PR.

Mixing many features together in one PR typically leads to issues down the road when something isn't working and makes everybody's work more difficult.

jianan-gu replied:
Totally understood, thanks for the suggestions, will do another two separate PRs for Torchscript and PyTorch SGD/Adagrad optimizer respectively.

jianan-gu replied:
PRs for [Torchscript integration](https://github.com/huggingface/transformers/pull/17153) and [SGD/Adagrad optimizer integration](https://github.com/huggingface/transformers/pull/17154) are summitted.
And also move the related codes out of this PR.
Besides, these three PRs have no strict dependency on the order of merging, while all the features mentioned in https://github.com/huggingface/transformers/issues/17137 will be landed once all of these PRs are merged.
Thank!

jianan-gu committed:
remove jit and optim bits

jianan-gu committed:
Merge branch 'huggingface:main' into Introduce_IPEX

jianan-gu committed:
Merge branch 'main' into Introduce_IPEX

jianan-gu committed:
Merge branch 'main' into Introduce_IPEX

jianan-gu committed:
black preview format

jianan-gu commented:
Hi,  @mfuntowicz @LysandreJik  @sgugger , any suggestion about this PR? We are happy to have a discussion here on any concerns.

Currently, BF16 AMP is already supported for GPU in Transformers, and with this PR we could extend that on the CPU side both for inference and training. Furtherly, users could get performance benefits by using IPEX with just one args "--use_ipex".

Thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -156,10 +158,15 @@
 if is_apex_available():
     from apex import amp
 
+if is_ipex_available() and "--use_ipex" in sys.argv:
+    import intel_extension_for_pytorch as ipex
```
We can't rely on `sys.argv` as the arguments may be passed to the `Trainer` class as `TrainingArguments`. This import should be done inside the `Trainer` as needed.

jianan-gu replied:
Got it, and have moved this importation to the inside `ipex_optimize_model`

sgugger commented on src/transformers/trainer.py:
```diff
@@ -156,10 +158,15 @@
 if version.parse(torch.__version__) >= version.parse("1.6"):
     _is_torch_generator_available = True
-    _is_native_amp_available = True
-    from torch.cuda.amp import autocast
```
Why is this import removed?

jianan-gu replied:
Because there are two kinds of autocast, for CPU and CUDA. Since torch is imported, we directly use `torch.cpu.amp.autocast` and `torch.cuda.amp.autocast` in contextmgr.

sgugger replied:
Thanks for explaining!

sgugger commented on src/transformers/trainer.py:
```diff
@@ -493,19 +501,30 @@ def __init__(
-                else:
-                    if args.bf16:
-                        raise ValueError("Tried to use `bf16` but native amp is not available")
+                if args.device == torch.device("cuda"):
```
This test is completely unreliable and will be false if `args.device` is set with `torch.device(0)` or when there are multiple GPUs. It should therefore be removed.

jianan-gu replied:
Got it, have made changes to check if it is CPU or others ( GPU, multiple GPU...), thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1139,7 +1167,23 @@ def call_model_init(self, trial=None):
         return model
 
+    def ipex_optimize_model(self, model, training=False, dtype=torch.float32):
+        if not training:
```
Looks like the `ipex` import can go here.

jianan-gu replied:
Have made changes accordingly, thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1139,7 +1167,23 @@ def call_model_init(self, trial=None):
+
     def _wrap_model(self, model, training=True):
+        if self.args.use_ipex:
+            if self.use_cpu_amp:
+                model = self.ipex_optimize_model(model, training, dtype=torch.bfloat16)
+            else:
+                model = self.ipex_optimize_model(model, training, dtype=torch.float32)
```
```suggestion
            dtype = torch.bfloat16 if self.use_cpu_amp else torch.float32
            model = self.ipex_optimize_model(model, training, dtype=dtype)
```

sgugger reviewed (commented):
Thanks for working on this. Before merging this we would need to have some tests of the feature. I also have a few comments on the PR.

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
refine code

jianan-gu committed:
refine notes

jianan-gu committed:
Merge branch 'huggingface:main' into Introduce_IPEX

sgugger commented on src/transformers/trainer.py:
```diff
@@ -494,19 +499,29 @@ def __init__(
+                if args.device == torch.device("cpu"):
+                    if args.fp16:
+                        raise ValueError("Tried to use `fp16` but it is not supported on cpu")
+                    else:
+                        if _is_native_cpu_amp_available:
+                            args.half_precision_backend = "cpu_amp"
+                        else:
+                            raise ValueError("Tried to use cpu amp but native cpu amp is not available")
```
```suggestion
                    elif _is_native_cpu_amp_available:
                        args.half_precision_backend = "cpu_amp"
                    else:
                        raise ValueError("Tried to use cpu amp but native cpu amp is not available")
```

sgugger commented on src/transformers/trainer.py:
```diff
@@ -494,19 +499,29 @@ def __init__(
+                        args.half_precision_backend = "cuda_amp"
                     else:
-                        args.half_precision_backend = "apex"
+                        if args.bf16:
+                            raise ValueError("Tried to use `bf16` but native amp is not available")
+                        else:
+                            args.half_precision_backend = "apex"
```
```suggestion
                    elif args.bf16:
                        raise ValueError("Tried to use `bf16` but native amp is not available")
                    else:
                        args.half_precision_backend = "apex"
```

sgugger reviewed (commented):
Thanks for addressing the comments! We're just missing a test now.

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
code refine

jianan-gu committed:
add ipex ut

jianan-gu commented:
> Thanks for addressing the comments! We're just missing a test now.

@sgugger Hi, here we have add ipex UT tests for trainer ( covering train, evaluate, predict). Could you please have a review of them? Thanks!

sgugger reviewed (approved):
Thanks for adding the tests. Looks good on my side, are you happy with the changes @stas00 ?

stas00 committed:
add performance cpu doc

stas00 reviewed (changes requested):
1. we need documentation otherwise nobody will know we have this new feature supported.

Since we didn't have a CPU-specific doc yet, I created a new doc `docs/source/en/perf_train_cpu.mdx`, @jianan-gu could you please fill in the details - in particular how a user will use ipex with HF Trainer. 

For reference of how to write this doc you can see for example this doc:
https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_train_gpu_one.mdx

I'd recommend adding at least one benchmark w/ and w/o IPEX to pique user's interest. Add `--skip_memory_metrics 0` to the Trainer args to get it prints speed stats for you.

2. I run the tests with pt-1.10 and pt-1.11

The problem is that installing
```
pip install intel_extension_for_pytorch
```
installs the wrong (latest pt) version and then the tests hang. 

For pt-1.10 one has to install
```
pip install intel_extension_for_pytorch==1.10.100+cpu -f https://software.intel.com/ipex-whl-stable
```

so I already added this nuance to the docs I have just created. The IPEX repo doesn't document that.

3.  We need to facilitate CI testing.

@ydshieh, do you know if our CI is Intel or AMD cpu? IPEX requires intel cpu to run. Thank you!

ydshieh commented:
I see `Intel Haswell` for the 4 runners (push/scheduled, single/multi GPUs)

stas00 commented:
> I see `Intel Haswell` for the 4 runners (push/scheduled, single/multi GPUs)

Great! Thank you, @ydshieh 

@sgugger, do we want the IPEX tests in live CI - they are very fast, but installing `intel_extension_for_pytorch` was very slow for pt-1.10 (was fast for pt-1.11) - or just scheduled?

stas00 committed:
Merge remote-tracking branch 'origin/main' into Introduce_IPEX

stas00 committed:
link to the cpu doc from main perf doc

stas00 committed:
install ipex into CI's docker

jianan-gu committed:
Update perf_train_cpu.mdx

stas00 commented on docs/source/en/perf_train_cpu.mdx:
```diff
@@ -15,23 +15,45 @@ This guide focuses on training large models efficiently on CPU.
 ## Mixed precision with IPEX
 
-XXX: Explain what it is - if you have an Intel CPU ...
+Low precision data type BFloat16 has been natively supported on the 3rd Generation Xeon® Scalable Processors (aka Cooper Lake) with AVX512 instruction set and will be supported on the next generation of Intel® Xeon® Scalable Processors with Intel® Advanced Matrix Extensions (Intel® AMX) instruction set with further boosted performance. The Auto Mixed Precision for CPU backend has been enabled sin ...
```
```suggestion
Low precision data type BFloat16 has been natively supported on the 3rd Generation Xeon® Scalable Processors (aka Cooper Lake) with AVX512 instruction set and will be supported on the next generation of Intel® Xeon® Scalable Processors with Intel® Advanced Matrix Extensions (Intel® AMX) instruction set with further boosted performance. The Auto Mixed Precision for CPU backend has been enabled since PyTorch-1.10. At the same time, the support of Auto Mixed Precision with BFloat16 for CPU and BFloat16 optimization of operators has been massively enabled in Intel® Extension for PyTorch, and partially upstreamed to PyTorch master branch. Users can get better performance and user experience with IPEX Auto Mixed Precision.
```

stas00 reviewed (commented):
The doc looks great - thank you, @jianan-gu!

Added one small grammar fix

Would you like to add a benchmark to show the speed with and without ipex? otherwise it's very hard to tell whether it's worth the effort trying it.

as I mentioned earlier you'd just run the trainer twice w/ and w/o ipex and make sure to add `--skip_memory_metrics 0` and it'll print you the speed at which it finished each run.

If it's not too much trouble that is. I'm asking since you probably have done a lot of experiments already and saw the power of this extension, while none of us did.

jianan-gu committed:
Update docs/source/en/perf_train_cpu.mdx

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

jianan-gu commented:
> The doc looks great - thank you, @jianan-gu!
> 
> Added one small grammar fix
> 
> Would you like to add a benchmark to show the speed with and without ipex? otherwise it's very hard to tell whether it's worth the effort trying it.
> 
> as I mentioned earlier you'd just run the trainer twice w/ and w/o ipex and make sure to add `--skip_memory_metrics 0` and it'll print you the speed at which it finished each run.
> 
> If it's not too much trouble that is. I'm asking since you probably have done a lot of experiments already and saw the power of this extension, while none of us did.

Hi, thanks for your suggestions and reply.
For the benchmark that shows the speed with and without ipex, yes, we could do that and present the speedup results there.
Also, we have done a lot of experiments like Bert training, and we will provide the performance gains but it would take some time (may cost 1-2 weeks) to go through the internal process to make them publicly available.
Besides, is it good enough if we demonstrate the speedup with relative performance gain (like the table [here](https://intel.github.io/intel-extension-for-pytorch/1.11.200/tutorials/performance.html))? Thanks

stas00 commented:
> Besides, is it good enough if we demonstrate the speedup with relative performance gain (like the table [here](https://intel.github.io/intel-extension-for-pytorch/1.11.200/tutorials/performance.html))?

I don't think so. Those performance numbers can't be reproduced by a user, so they are very practical. 

As I suggested there is no need for any major benchmarks, here is a simple eval one:

```
# no gpus / no ipex 
PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES= python \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-small \
--output_dir output_dir --adam_eps 1e-06 --do_eval --evaluation_strategy=steps \
--label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step \
--logging_steps 500 --max_source_length 128 --max_target_length 128 \
--num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size 16 \
--predict_with_generate --eval_steps 150 --sortish_sampler --source_lang en \
--target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix \
'translate English to Romanian: ' --val_max_target_length 128 --warmup_steps \
50 --max_eval_samples 500 --skip_memory_metrics 0 --bf16
[...]
***** eval metrics *****
  before_init_mem_cpu       =     1000MB
  eval_bleu                 =    24.1261
  eval_gen_len              =     39.554
  eval_loss                 =      3.786
  eval_mem_cpu_alloc_delta  =      279MB
  eval_mem_cpu_peaked_delta =      437MB
  eval_runtime              = 0:01:53.04
  eval_samples              =        500
  eval_samples_per_second   =      4.423
  eval_steps_per_second     =      0.283
  init_mem_cpu_alloc_delta  =        0MB
  init_mem_cpu_peaked_delta =        0MB
  
# no gpus / no ipex 
PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES= python \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-small \
--output_dir output_dir --adam_eps 1e-06 --do_eval --evaluation_strategy=steps \
--label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step \
--logging_steps 500 --max_source_length 128 --max_target_length 128 \
--num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size 16 \
--predict_with_generate --eval_steps 150 --sortish_sampler --source_lang en \
--target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix \
'translate English to Romanian: ' --val_max_target_length 128 --warmup_steps \
50 --max_eval_samples 500 --skip_memory_metrics 0 --bf16 --no_cuda --use_ipex
***** eval metrics *****
  before_init_mem_cpu       =     1010MB
  eval_bleu                 =    24.1261
  eval_gen_len              =     39.554
  eval_loss                 =     3.7863
  eval_mem_cpu_alloc_delta  =      518MB
  eval_mem_cpu_peaked_delta =      439MB
  eval_runtime              = 0:01:34.41
  eval_samples              =        500
  eval_samples_per_second   =      5.296
  eval_steps_per_second     =      0.339
  init_mem_cpu_alloc_delta  =        0MB
  init_mem_cpu_peaked_delta =        0MB

```

So 5.296 vs  4.423 eval_samples_per_second on my machine. ~16% speedup - that's a good start.

```
model name      : 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
```

Probably would be better to test with a slightly larger model, as it is likely to get better speedup but it helps to see that it actually works!

This is good enough for a doc.

And down the road we can surely boost it with better more indepth benchmarks.

What do you think?

stas00 commented:
> Low precision data type BFloat16 has been natively supported on the 3rd Generation Xeon® Scalable Processors (aka Cooper Lake) with AVX512 instruction set and will be supported on the next generation of Intel® Xeon® Scalable Processors with Intel® Advanced Matrix Extensions (Intel® AMX) instruction set with further boosted performance.

Given your description - is there a way to check that ipex is actually doing anything, other than being installed and import'able?

i.e. currently it's just:

```
def is_ipex_available():
    return importlib.util.find_spec("intel_extension_for_pytorch") is not None
```

But if user's CPU is not Intel or not the right kind of Intel should it tell the user it's not supported and assert? or is it already the case? I guess mine is the right CPU so I it doesn't fail.

jgong5 commented:
> But if user's CPU is not Intel or not the right kind of Intel should it tell the user it's not supported and assert? or is it already the case? I guess mine is the right CPU so I it doesn't fail.

Hi @stas00 The optimization in IPEX is a superset of BF16. It optimizes for CPU with AVX-512 or above. It can also functionally work for CPUs with AVX2. So, I expect it also functionally works for AMD CPUs (even though we do not benchmark perf on AMD CPUs). IPEX doesn't explicitly check CPU types internally. With this, we are open to whether and what additional checks are needed to add in this PR.

jianan-gu commented:
> > Besides, is it good enough if we demonstrate the speedup with relative performance gain (like the table [here](https://intel.github.io/intel-extension-for-pytorch/1.11.200/tutorials/performance.html))?
> 
> I don't think so. Those performance numbers can't be reproduced by a user, so they are very practical.
> 
> As I suggested there is no need for any major benchmarks, here is a simple eval one:
> 
> ```
> # no gpus / no ipex 
> PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES= python \
> examples/pytorch/translation/run_translation.py --model_name_or_path t5-small \
> --output_dir output_dir --adam_eps 1e-06 --do_eval --evaluation_strategy=steps \
> --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step \
> --logging_steps 500 --max_source_length 128 --max_target_length 128 \
> --num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size 16 \
> --predict_with_generate --eval_steps 150 --sortish_sampler --source_lang en \
> --target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix \
> 'translate English to Romanian: ' --val_max_target_length 128 --warmup_steps \
> 50 --max_eval_samples 500 --skip_memory_metrics 0 --bf16
> [...]
> ***** eval metrics *****
>   before_init_mem_cpu       =     1000MB
>   eval_bleu                 =    24.1261
>   eval_gen_len              =     39.554
>   eval_loss                 =      3.786
>   eval_mem_cpu_alloc_delta  =      279MB
>   eval_mem_cpu_peaked_delta =      437MB
>   eval_runtime              = 0:01:53.04
>   eval_samples              =        500
>   eval_samples_per_second   =      4.423
>   eval_steps_per_second     =      0.283
>   init_mem_cpu_alloc_delta  =        0MB
>   init_mem_cpu_peaked_delta =        0MB
>   
> # no gpus / no ipex 
> PYTHONPATH=src USE_TF=0 CUDA_VISIBLE_DEVICES= python \
> examples/pytorch/translation/run_translation.py --model_name_or_path t5-small \
> --output_dir output_dir --adam_eps 1e-06 --do_eval --evaluation_strategy=steps \
> --label_smoothing 0.1 --learning_rate 3e-5 --logging_first_step \
> --logging_steps 500 --max_source_length 128 --max_target_length 128 \
> --num_train_epochs 1 --overwrite_output_dir --per_device_eval_batch_size 16 \
> --predict_with_generate --eval_steps 150 --sortish_sampler --source_lang en \
> --target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix \
> 'translate English to Romanian: ' --val_max_target_length 128 --warmup_steps \
> 50 --max_eval_samples 500 --skip_memory_metrics 0 --bf16 --no_cuda --use_ipex
> ***** eval metrics *****
>   before_init_mem_cpu       =     1010MB
>   eval_bleu                 =    24.1261
>   eval_gen_len              =     39.554
>   eval_loss                 =     3.7863
>   eval_mem_cpu_alloc_delta  =      518MB
>   eval_mem_cpu_peaked_delta =      439MB
>   eval_runtime              = 0:01:34.41
>   eval_samples              =        500
>   eval_samples_per_second   =      5.296
>   eval_steps_per_second     =      0.339
>   init_mem_cpu_alloc_delta  =        0MB
>   init_mem_cpu_peaked_delta =        0MB
> ```
> 
> So 5.296 vs 4.423 eval_samples_per_second on my machine. ~16% speedup - that's a good start.
> 
> ```
> model name      : 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
> ```
> 
> Probably would be better to test with a slightly larger model, as it is likely to get better speedup but it helps to see that it actually works!
> 
> This is good enough for a doc.
> 
> And down the road we can surely boost it with better more indepth benchmarks.
> 
> What do you think?

Sure, got it, so we will prepare for benchmarking example models to collect performance numbers (like you showed above) and the internal review process. And after that, we will update the doc with those performance numbers.
Thanks.

stas00 commented:
>      Hi @stas00 The optimization in IPEX is a superset of BF16. It optimizes for CPU with AVX-512 or above. It can also functionally work for CPUs with AVX2. So, I expect it also functionally works for AMD CPUs (even though we do not benchmark perf on AMD CPUs). IPEX pre-compiles multiple kernels for various CPU ISAs for an op ahead-of-time and does runtime kernel dispatch according to the underlying CPU ISA capability. IPEX doesn't explicitly check CPU types internally. With this, we are open to whether and what additional checks are needed to add in this PR.

Perhaps mention it in the doc? just briefly - as in AMD CPUs and older Intel CPUs are likely to result in better performance as well

> Sure, got it, so we will prepare for benchmarking example models to collect performance numbers (like you showed above) and the internal review process. And after that, we will update the doc with those performance numbers.

I'd be perfectly happy to merge this sooner not to press you if you kindly commit to adding at least some benchmarks later - will that be a good arrangement for you guys?

stas00 committed:
Merge remote-tracking branch 'origin/main' into Introduce_IPEX

jianan-gu committed:
Update perf_train_cpu.mdx

jianan-gu committed:
Update perf_train_cpu.mdx

jgong5 commented:
Hi @stas00 
> Perhaps mention it in the doc? just briefly - as in AMD CPUs and older Intel CPUs are likely to result in a better performance as well under IPEX.

Thanks for the suggestions. We revised the doc accordingly. Could you please review?

> I'd be perfectly happy to merge this sooner not to press you if you kindly commit to adding at least some benchmarks later in a new PR - will that be a good arrangement for you guys?

Sure, sounds good to us. We will update the perf numbers in a separate PR as soon as they are ready.

stas00 reviewed (approved):
Looks good now, @jianan-gu 

@sgugger - could you please have another look - just the docs and then it's good to merge. Thank you!

sgugger reviewed (approved):
LGTM, thanks a lot for iterating on this!

stas00 commented:
@jianan-gu, fyi, ipex fails with pt-1.12

```
tests/trainer/test_trainer.py::TrainerIntegrationTest::test_evaluate_with_ipex
(line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
tests/trainer/test_trainer.py::TrainerIntegrationTest::test_number_of_steps_in_training_with_ipex
(line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
tests/trainer/test_trainer.py::TrainerIntegrationTest::test_predict_with_ipex
(line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
```

we are disabling the tests for now, and please let us know when this is fixed, so that we can reenable those. Thank you!

jianan-gu commented:
> @jianan-gu, fyi, ipex fails with pt-1.12
> 
> ```
> tests/trainer/test_trainer.py::TrainerIntegrationTest::test_evaluate_with_ipex
> (line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
> tests/trainer/test_trainer.py::TrainerIntegrationTest::test_number_of_steps_in_training_with_ipex
> (line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
> tests/trainer/test_trainer.py::TrainerIntegrationTest::test_predict_with_ipex
> (line 6)  ImportError: /usr/local/lib/python3.8/dist-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-cpu.so: undefined symbol: _ZNK3c1010TensorImpl5sizesEv
> ```
> 
> we are disabling the tests for now, and please let us know when this is fixed, so that we can reenable those. Thank you!

Hi @stas00 
Thanks for the information.
As mentioned in this issue, IPEX 1.12 release is [available ](https://intel.github.io/intel-extension-for-pytorch/1.12.0/tutorials/installation.html), and we also open a PR https://github.com/huggingface/transformers/pull/18072 to enhance the integration for this version mismatch issue to avoid breaking Trainer;

Besides, for the performance data we discussed in this PR, since we were mostly working on the IPEX 1.12 release during the past weeks, we would like to prepare the data based on the new release. It would need more time, thanks.

Oxi84 commented:
I do not understand where is this implementation added?
    
    import transformers
    from transformers import T5ForConditionalGeneration,T5Tokenizer,T5TokenizerFast
    model1a = T5ForConditionalGeneration.from_pretrained('t5-base',low_cpu_mem_usage=True)
    tokenizer1 = T5TokenizerFast.from_pretrained('t5-base')

stas00 commented:
@Oxi84, please see: https://huggingface.co/docs/transformers/perf_train_cpu#mixed-precision-with-ipex
